Update migration doc (#20)
ably/laravel-broadcaster: README.md

<a name="migrate-pusher-to-ably"></a>
## Migrating from pusher/pusher-compatible broadcasters

> The current Ably broadcaster is fully compatible with the [pusher](https://laravel.com/docs/9.x/broadcasting#pusher-channels), [old Ably Broadcaster](https://laravel.com/docs/9.x/broadcasting#ably) and [pusher compatible open source broadcasters](https://laravel.com/docs/9.x/broadcasting#open-source-alternatives). Please follow below steps to migrate properly from other broadcasters.

Author: done!

> **1. Leaving the channel**

Author: I think in terms of `echo` terminology, it's better to call it `leaving the presence channel`

> **1. Leaving the channel**

Author: Hey, sorry this section is not about leaving channel presence, rather leaving the channel itself. In terms of ably terminology it's `unsubscribing to the channel`. 
So, it's possible with public, private or presence channels.

Leaving the channel (echo) = unsubscribe to channel (ably)

> **1. Leaving the channel**

Reviewer: Ah I see, thanks for explaining 👍

In that case we should change it to this so that it matches the [laravel documentation](https://laravel.com/docs/9.x/broadcasting#leaving-a-channel)
```suggestion
**1. Leaving a channel**
```

> For Leaving the channel on client side, you should use [Ably Channel Namespaces](https://ably.com/docs/general/channel-rules-namespaces) conventions.

Author: done!

Echo.join('channel3'); 
Echo.leaveChannel("presence:channel3")
```

> instead of [Pusher Channel Conventions](https://pusher.com/docs/channels/using_channels/channels/#channel-types).

Author: This is confusing, in a sense how channels are created in the first place. 
They don't have a prefix, yet this example can put you into a dilemma whether the channel was created with the given prefix.
So, we also need to provide how we create a channel

```
 // public channel
Echo.channel('channel1');
Echo.leaveChannel("public:channel1");
// private channel
Echo.private('channel2'); 
Echo.leaveChannel("private:channel2")
// presence channel
Echo.join('channel3'); 
Echo.leaveChannel("presence:channel3")
```

I am not in favor of mixing `pusher` conventions with `ably`, it's like mixing wrong and right code together.

> instead of [Pusher Channel Conventions](https://pusher.com/docs/channels/using_channels/channels/#channel-types).

Author: Also, I am not in favor of mixing `pusher` conventions with `ably`, it's like mixing wrong and right code together.

> instead of [Pusher Channel Conventions](https://pusher.com/docs/channels/using_channels/channels/#channel-types).

Author: You can check official doc => 
https://laravel.com/docs/9.x/broadcasting#leaving-a-channel

This information is not enough at all. Leaving the channel section is confusing, we are trying to simplify it for both `pusher` and `ably`.

> instead of [Pusher Channel Conventions](https://pusher.com/docs/channels/using_channels/channels/#channel-types).

Reviewer: ok that makes sense thanks. I still think having two large snippets to explain this is quite awkward to read. how about this?

```js
 // For public channels
Echo.channel("channel1");
// Use this:
Echo.leaveChannel("public:channel1");
// Instead of:
Echo.leaveChannel("channel1");

// etc...
```

> instead of [Pusher Channel Conventions](https://pusher.com/docs/channels/using_channels/channels/#channel-types).

Author: done

Echo.leaveChannel("presence-channel3")
```

**2. Error handling**
- [Ably echo client](https://github.com/ably-forks/laravel-echo) emit [ably specific error codes](https://github.com/ably/ably-common/blob/main/protocol/errors.json) instead of [pusher error codes](https://pusher.com/docs/channels/library_auth_reference/pusher-websockets-protocol/#error-codes).

> - Aim is to make sure error details are descriptive and easy to understand, so it's more effective to take a corrective action ( Pusher errors lack error context, mostly emitted as integer error codes ).

Reviewer: I think remove this line, it seems a bit dismissive of Pusher SDKs and doesn't add much value

Author: I think it should be dismissive. 
I would say error details add much more value rather than just emitting int error code. 
Pusher SDKs truly suck at handling errors. It's hard to find out what's going wrong. This was one of the reasons why lot of people used to ask questions about `ably-pusher` setup, and it does take a considerable amount of time to fix it. I personally faced tons of issues while the initial setup.
I can confidently say, this is true for all pusher SDK users.

Author: Okay I have updated note to =>
```
Ably emitted errors are descriptive and easy to understand, so it's more effective to take a corrective action ( Pusher errors mostly emitted as integer error codes ).
```
Now, it's not dismissive and mentions that ably errors are more informative.

Reviewer: That's better but I would still prefer to remove the "( Pusher errors mostly emitted as integer error codes )" bit tbh

Author: done

> - Ably errors are provided as a [errorInfo object](https://ably.com/docs/api/realtime-sdk/types#error-info) with proper error context.

Reviewer: ```suggestion
- Ably errors are provided as an [ErrorInfo object](https://ably.com/docs/api/realtime-sdk/types#error-info) with full error context.
```

> - Care needs to be taken while checking on the errorInfo object and mapping needs to be done from [pusher error codes](https://pusher.com/docs/channels/library_auth_reference/pusher-websockets-protocol/#error-codes) to [ably error codes](https://github.com/ably/ably-common/blob/main/protocol/errors.json).

Reviewer: ```suggestion
- If you are interacting with Pusher errors in your project, be sure to update your code accordingly.
```

Author: done!

```js
    channel.error(error => {
        if (error && error.code === 40142) { // ably token expired
            console.error(error);
            // take corrective action on UI
        }
    })
```
**Note :**
- AblyPresenceChannel -> `here` method gets called everytime client **joins, updates or leaves** the channel unlike pusher client (gets called only for the first time).

> - This is perfectly in sync with interface method [**here** provided under **presence-channel.ts**](https://github.com/laravel/echo/blob/master/src/channel/presence-channel.ts#L10).

Reviewer: I'm not entirely sure what this sentence means, could you clarify for me pls?

Author: `here` method for pusher gets called only for the first time. This is supposed to be called whenever user joins, updates or leaves the presence as per echo interface.

Reviewer: yeah i get line 178, i'm asking about line 179 though

Author: It means that it follows the implementation as per the standard PresenceChannel interface.

Author: Also, means `pusher` is doing it wrong : )

Reviewer: Oh I see, that makes sense, I think you should remove this line then, it's quite confusing and doesn't seem necessary

Author: I have remove the line, and updated it with line that will actually make sense.

```

**2. Error handling**
- Please note that the [Ably laravel-echo client](https://github.com/ably-forks/laravel-echo) emits [Ably specific error codes](https://github.com/ably/ably-common/blob/main/protocol/errors.json) instead of [Pusher error codes](https://pusher.com/docs/channels/library_auth_reference/pusher-websockets-protocol/#error-codes).
- Ably emitted errors are descriptive and easy to understand, so it's more effective to take a corrective action ( Pusher errors mostly emitted as integer error codes ).
- Ably errors are provided as an [ErrorInfo object](https://ably.com/docs/api/realtime-sdk/types#error-info) with full error context.
- If you are interacting with pusher errors in your project, be sure to update your code accordingly.
i.e. update from [pusher error codes](https://pusher.com/docs/channels/library_auth_reference/pusher-websockets-protocol/#error-codes) to [ably error codes](https://github.com/ably/ably-common/blob/main/protocol/errors.json).

```js
    channel.error(error => {
        if (error && error.code === 40142) { // ably token expired
            console.error(error);
            // take corrective action on UI
        }
    })
```
**Note :**

> - AblyPresenceChannel -> `here` method gets called everytime client **joins, updates or leaves** the channel unlike pusher client (gets called only for the first time).

Reviewer: ```suggestion
- AblyPresenceChannel -> `here` method gets called every time a client **joins, updates or leaves** the channel, whereas when using Pusher this is only called when a client enters.
```

Author: It's not when `client` enters. It's called only once in the whole lifetime for the first client entering the channel.

Author: updated!
